common/src/main/java/com/wynntils/features/user/map/MinimapFeature.java

```diff
@@ -218,31 +218,35 @@ private void renderPois(
                 double playerZ,
                 BoundingBox textureBoundingBox) {
 
-            float sinRotationRadians = 0f;
-            float cosRotationRadians = 0f;
+            float sinRotationRadians;
+            float cosRotationRadians;
 
             if (followPlayerRotation) {
                 double rotationRadians = Math.toRadians(McUtils.player().getYRot());
                 sinRotationRadians = (float) StrictMath.sin(rotationRadians);
                 cosRotationRadians = (float) -StrictMath.cos(rotationRadians);
+            } else {
+                sinRotationRadians = 0f;
+                cosRotationRadians = 0f;
             }
 
             float currentZoom = 1f / scale;
 
-            List<Poi> poisToRender = new ArrayList<>(Models.Map.getServicePois());
-            poisToRender.addAll(MapFeature.INSTANCE.customPois);
-            List<PlayerMiniMapPoi> playerPois = Models.HadesUser.getHadesUserMap().values().stream()
-                    .filter(user -> (user.isPartyMember() && renderRemotePartyPlayers)
-                            || (user.isMutualFriend() && renderRemoteFriendPlayers))
-                    .map(PlayerMiniMapPoi::new)
-                    .toList();
-            poisToRender.addAll(playerPois);
+            Stream<? extends Poi> poisToRender = Models.Map.getServicePois().stream();
+
+            poisToRender = Stream.concat(
+                    poisToRender,
+                    Models.HadesUser.getHadesUserMap().values().stream()
+                            .filter(user -> (user.isPartyMember() && renderRemotePartyPlayers)
+                                    || (user.isMutualFriend() && renderRemoteFriendPlayers))
+                            .map(PlayerMiniMapPoi::new));
 
-            poisToRender.addAll(Models.Map.getCombatPois());
+            poisToRender = Stream.concat(poisToRender, Models.Map.getCombatPois().stream());
 
             // Reverse order to make sure higher priority is drawn later than lower priority to overwrite them
-            poisToRender.sort(Comparator.comparing(Poi::getDisplayPriority).reversed());
-            for (Poi poi : poisToRender) {
+            poisToRender = poisToRender.sorted(
```

Comment:
In my measurements, the sort took a fair bit of time as well. 
Do you think we could map display priority to Z value so we could use the rendering system to make sure more prioritized icons render on top of less important, instead of sorting this for each frame?

That's a separate optimization, though, so I'll accept this as the arraylist stuff.

Reply:
Actually, I've not tried batching and render Y levels. Can you add this into the list of optimizations and I will take a look? I have a feeling this can work.

Reply:
I think I said something about the sorting, but I can make sure it is more explicit. The idea with the Z value did not strike me until just right now.